### PR TITLE
optimize CPU implementation of Attention

### PR DESCRIPTION
**Pull request by fs-eire**
**Description**: This change optimizes Attention to use parallelization to improve compute latency.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/bert/attention.cc [hunk 1]
```diff
@@ -115,151 +115,175 @@ Status Attention<T>::Compute(OpKernelContext* context) const {
...
-  ORT_RETURN_IF_ERROR(TransposeBase::DoTranspose(transpose_permutations, gemm_data_tensor, gemm_data_transposed_tensor));
-
-  T* Q = reinterpret_cast<T*>(gemm_data_transposed);
-  T* K = Q + (batch_size * hidden_size * sequence_length);
-  T* V = K + (batch_size * hidden_size * sequence_length);
-
-  // STEP.3: scratch(B, N, S, S) = 1/sqrt(H) x Q(B, N, S, H) x K'(B, N, S, H -> B, N, H, S) + 1 x mask_index(B -> B, 1, 1, 1)
-  auto scratch_data = allocator->Alloc(batch_size * num_heads_ * sequence_length * sequence_length * element_size);
-  BufferUniquePtr scratch_buffer(scratch_data, BufferDeleter(allocator));
+  T* QKV[3] = {reinterpret_cast<T*>(Q), reinterpret_cast<T*>(K), reinterpret_cast<T*>(V)};
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/bert/attention.cc]: reinterpret_cast is needless here

File: onnxruntime/contrib_ops/cpu/bert/attention.cc [hunk 2]
```diff
@@ -115,151 +115,175 @@ Status Attention<T>::Compute(OpKernelContext* context) const {
...
-            head_size,
-            1.0f / sqrt(static_cast<float>(head_size)),
-            Q + offset_Q,
-            K + offset_Q,
-            1.0,
-            reinterpret_cast<T*>(scratch_data) + offset_scratch,
-            tp);
-        offset_Q += offset_Q_increment;
-        offset_scratch += offset_scratch_increment;
+      int mask = mask_index->template Data<int>()[b_i];
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/attention.cc]: int32_t

Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/bert/attention.cc]: mask_index can be used in softmax to save some calculation. See cuda kernel usage. Just a hint for further improvement in the future.

File: onnxruntime/contrib_ops/cpu/bert/attention.cc [hunk 3]
```diff
@@ -115,151 +115,175 @@ Status Attention<T>::Compute(OpKernelContext* context) const {
...
+          1.0,
+          reinterpret_cast<T*>(scratch_data) + sequence_length * sequence_length * i,
+          nullptr);
+    });
+  }
 
+  // STEP.3: P(B, N, S, S) = Softmax(scratch)
   {
     int N = batch_size * num_heads_ * sequence_length;
     int D = sequence_length;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/attention.cc]: const for N and D

File: onnxruntime/contrib_ops/cpu/bert/attention.cc [hunk 4]
```diff
@@ -115,151 +115,175 @@ Status Attention<T>::Compute(OpKernelContext* context) const {
...
   ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&allocator));
 
-  concurrency::ThreadPool* tp = context->GetOperatorThreadPool();
-
-  // STEP.1: gemm_data(BS, 3NH) = input(BS, NH) x weights(NH, 3NH) + bias(NH)
+  // STEP.1: gemm_data(BS, 3NH) = input(BS, NH) x weights(NH, 3NH) + bias(3NH)
   auto gemm_data = allocator->Alloc(batch_size * sequence_length * 3 * hidden_size * element_size);
   BufferUniquePtr gemm_buffer(gemm_data, BufferDeleter(allocator));
+  auto Q = reinterpret_cast<T*>(gemm_data);
+  auto K = Q + batch_size * sequence_length * hidden_size;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/attention.cc]: store batch_size * sequence_length * hidden_size to a variable. It is used in multiple places in this function.

File: onnxruntime/contrib_ops/cpu/bert/attention.cc (hunk 2 above)
Comment (fs-eire) [reply onnxruntime/contrib_ops/cpu/bert/attention.cc]: I added comments to make this as a TODO item

**Pull request merged**